src/components/templates/PartyMap/components/Map/Map.tsx

```diff
@@ -73,65 +100,80 @@ export const Map: React.FC<PropsType> = ({
       const calcRows = venue.columns
         ? Math.round(parseInt(venue.columns.toString()) / imgRatio)
         : DEFAULT_ROWS;
-      setRows(calcRows);
+      setRowsArray(Array.from(Array(calcRows)));
     };
   }, [venue.columns, venue.mapBackgroundImageUrl]);
 
-  const roomsHit = useMemo(() => {
-    if (!currentPosition?.row || !currentPosition?.column) return [];
-
-    const { row, column } = currentPosition;
-
-    const roomHitFilter = makeCampRoomHitFilter({
-      row,
-      column,
-      totalRows: rows,
-      totalColumns: columns,
-    });
+  const prevSeat = useRef<{
+    row: number | null;
+    column: number | null;
+  }>();
 
-    return venue.rooms?.filter(roomHitFilter);
-  }, [currentPosition, rows, columns, venue.rooms]);
-
-  const detectRoomsOnMove = useCallback(() => {
-    if (selectedRoom) {
-      const noRoomHits = hasElements(roomsHit);
-      if (!noRoomHits && selectedRoom) {
-        setSelectedRoom(undefined);
-        setIsRoomModalOpen(false);
+  const takeSeat = useCallback(
+    async (row: number | null, column: number | null) => {
+      if (!userUid || !venueId) return;
+
+      if (
+        prevSeat.current?.column === column &&
+        prevSeat.current?.row === row
+      ) {
+        return;
       }
-    }
+      prevSeat.current = { row, column };
 
-    roomsHit?.forEach((room) => {
-      setSelectedRoom(room);
-      setIsRoomModalOpen(true);
-    });
-  }, [roomsHit, selectedRoom, setIsRoomModalOpen, setSelectedRoom]);
+      const doc = `users/${userUid}`;
+      // FIXME: @debt This way it will rewrite all other locations.
+      // But if you turn it on, the performance will drop drastically,
+      // because profile will be updated on every function call
 
-  const takeSeat = useCallback(
-    (row: number | null, column: number | null) => {
-      if (!user || !profile || !venueId) return;
-      const doc = `users/${user.uid}`;
-      const existingData = profile?.data;
+      // const existingData = profile?.data;
       const update = {
         data: {
-          ...existingData,
+          // ...existingData,
           [venueId]: {
             row,
             column,
           },
         },
       };
-      const firestore = firebase.firestore();
-      firestore
+
+      await firestore
         .doc(doc)
         .update(update)
         .catch(() => {
           firestore.doc(doc).set(update);
         });
```

Comment:
I know this is a pattern we use in a lot of places currently, but this "try and `update`, then catch the error if it doesn't exist and just `set` it instead" is an anti-pattern/tech-debt in our codebase in my opinion.

I would have to check the firebase docs again to be sure, but I believe there is a function that just handles the 'upsert' pattern anyway, without resorting to this sort of hack. Or if not, we could check if the record exists up front, then choose whether to `update` or `insert` as appropriate.

Reply:
Agreed with your that this looks hacky. However, I'd change this in `refactoring` PR, changing all the occurrences of this across our codebase. Will leave it as is in this branch.

Reply:
backlogged in https://github.com/sparkletown/internal-sparkle-issues/issues/125